src/Test/Inspection/Core.hs

```diff
@@ -203,43 +204,49 @@ eqSlice it slice1 slice2
 -- | Returns @True@ if the given core expression mentions no type constructor
 -- anywhere that has the given name.
 freeOfType :: Slice -> [Name] -> Maybe (Var, CoreExpr)
-freeOfType slice tcNs = allTyCons (\tc -> getName tc `notElem` tcNs) slice
-
-allTyCons :: (TyCon -> Bool) -> Slice -> Maybe (Var, CoreExpr)
-allTyCons predicate slice = listToMaybe [ (v,e) | (v,e) <- slice, not (go e) ]
+freeOfType slice tcNs =
+      fmap (\(a,b,_) -> (a,b))
+    $ allTyCons (\tc -> getName tc `notElem` tcNs) slice
+
+-- | Check if all type constructors in a slice satisfy the given predicate.
+-- Returns the binder, expression and failing constructors triple on failure.
+allTyCons :: (TyCon -> Bool) -> Slice -> Maybe (Var, CoreExpr, [TyCon])
+allTyCons predicate slice =
+    listToMaybe
+        [(v, e, nub tcs) | (v, e) <- slice, let tcs = go e, not (null tcs)]
   where
     goV v = goT (varType v)
 
     go (Var v)           = goV v
-    go (Lit _ )          = True
-    go (App e a)         = go e && go a
-    go (Lam b e)         = goV b && go e
-    go (Let bind body)   = all goB (flattenBinds [bind]) && go body
-    go (Case s b _ alts) = go s && goV b && all goA alts
+    go (Lit _)           = []
+    go (App e a)         = go e ++ go a
+    go (Lam b e)         = goV b ++ go e
+    go (Let bind body)   = concatMap goB (flattenBinds [bind]) ++ go body
+    go (Case s b _ alts) = go s ++ goV b ++ concatMap goA alts
     go (Cast e _)        = go e
     go (Tick _ e)        = go e
     go (Type t)          = (goT t)
-    go (Coercion _)      = True
+    go (Coercion _)      = []
 
-    goB (b, e) = goV b && go e
+    goB (b, e) = goV b ++ go e
 
-    goA (_,pats, e) = all goV pats && go e
+    goA (_,pats, e) = concatMap goV pats >> go e
```

Comment:
I'm surprised to see `>>` here. Is it supposed to be `>>=`?

Reply:
Thanks for catching the bug! This should have been `++` instead of `>>`.  This should be the cause of #41 .

In the older version of the PR we were using `ExceptT` and `>>` was the right thing then. When it was changed to use `concatMap` `>>` was changed to `++`, this particular instance remained somehow.

Reply:
Well spotted! :-)